src/olympia/discovery/tests/test_admin.py

```diff
@@ -1038,3 +1039,190 @@ def test_need_3_modules(self):
+        self.grant_permission(user, 'Discovery:Edit')
+        self.client.login(email=user.email)
+        response = self.client.get(detail_url, follow=True)
+        assert response.status_code == 200
+        content = response.content.decode('utf-8')
+        assert 'Recommended extensions' in content
+
+        response = self.client.post(
+            detail_url,
+            {
+                'title': 'Popular extensions',
+                'shelfType': 'extension',
+                'criteria': '/this/is/the/criteria',
+                'footerText': 'See more',
+                'footerPathname': '/this/is/the/pathname'
+            }, follow=True)
+        assert response.status_code == 200
+        item.reload()
+        assert Shelf.objects.count() == 1
+        assert item.title == 'Popular extensions'
+
+    def test_can_delete_with_discovery_edit_permission(self):
+        item = Shelf.objects.create(
+            title='Recommended extensions',
+            shelfType='extension',
+            criteria='/this/is/the/criteria',
+            footerText='See more',
+            footerPathname='/this/is/the/pathname')
+        delete_url = reverse(
+            'admin:discovery_shelfmodule_delete', args=(item.pk,)
+        )
+        user = user_factory()
+        self.grant_permission(user, 'Admin:Tools')
+        self.grant_permission(user, 'Discovery:Edit')
+        self.client.login(email=user.email)
+        # Can access delete confirmation page.
+        response = self.client.get(delete_url, follow=True)
+        assert response.status_code == 200
+        assert Shelf.objects.filter(pk=item.pk).exists()
+
+        # And can actually delete.
+        response = self.client.post(
+            delete_url, data={'post': 'yes'}, follow=True)
+        assert response.status_code == 200
+        assert not Shelf.objects.filter(pk=item.pk).exists()
+
+    def test_can_add_with_discovery_edit_permission(self):
+        add_url = reverse('admin:discovery_shelfmodule_add')
+        user = user_factory()
+        self.grant_permission(user, 'Admin:Tools')
+        self.grant_permission(user, 'Discovery:Edit')
+        self.client.login(email=user.email)
+        response = self.client.get(add_url, follow=True)
+        assert response.status_code == 200
+        assert Shelf.objects.count() == 0
+        response = self.client.post(
+            add_url,
+            {
+                'title': 'Recommended extensions',
+                'shelfType': 'extension',
+                'criteria': '/this/is/the/criteria',
+                'footerText': 'See more',
+                'footerPathname': '/this/is/the/pathname'
+            },
+            follow=True)
+        assert response.status_code == 200
+        assert Shelf.objects.count() == 1
+        item = Shelf.objects.get()
+        assert item.title == 'Recommended extensions'
+
+    def test_can_not_add_with_discovery_edit_permission(self):
```

Comment:
Should this read "test_can_not_add_with*out*_discovery_edit_permission"?

Reply:
Yes, yes it should. Good eyes! Updated!